<a href="https://colab.research.google.com/github/DenisDrobyshev/university/blob/master/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22lesson1_practice_pandas_titanic_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ данных пассажиров "Титаника"</center>


In [1]:
import numpy as np
import pandas as pd

%matplotlib inline

**Считаем данные из файла в память в виде объекта Pandas.DataFrame**

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/refs/heads/master/titanic.csv")

**Данные представлены в виде таблицы. Посмотрим на первые 5 строк:**

In [4]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

**Для примера отберем пассажиров, которые сели в Cherbourg (Embarked=C) и заплатили более 200 у.е. за билет (fare > 200).**

In [5]:
df[(df['Embarked'] == 'C') & (df['Fare'] > 200)]

PassengerId  Survived  Pclass  \
118          119         0       1   
258          259         1       1   
299          300         1       1   
311          312         1       1   
377          378         0       1   
380          381         1       1   
557          558         0       1   
679          680         1       1   
700          701         1       1   
716          717         1       1   
737          738         1       1   
742          743         1       1   

                                                  Name     Sex   Age  SibSp  \
118                           Baxter, Mr. Quigg Edmond    male  24.0      0   
258                                   Ward, Miss. Anna  female  35.0      0   
299    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
311                         Ryerson, Miss. Emily Borie  female  18.0      2   
377                          Widener, Mr. Harry Elkins    male  27.0      0   
380                              Bidois, Miss. Rosalie  female  42.0      0   
557                                Robbins, Mr. Victor    male   NaN      0   
679                 Cardeza, Mr. Thomas Drake Martinez    male  36.0      0   
700  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0      1   
716                      Endres, Miss. Caroline Louise  female  38.0      0   
737                             Lesurer, Mr. Gustave J    male  35.0      0   
742              Ryerson, Miss. Susan Parker "Suzette"  female  21.0      2   

     Parch    Ticket      Fare            Cabin Embarked  
118      1  PC 17558  247.5208          B58 B60        C  
258      0  PC 17755  512.3292              NaN        C  
299      1  PC 17558  247.5208          B58 B60        C  
311      2  PC 17608  262.3750  B57 B59 B63 B66        C  
377      2    113503  211.5000              C82        C  
380      0  PC 17757  227.5250              NaN        C  
557      0  PC 17757  227.5250              NaN        C  
679      1  PC 17755  512.3292      B51 B53 B55        C  
700      0  PC 17757  227.5250          C62 C64        C  
716      0  PC 17757  227.5250              C45        C  
737      0  PC 17755  512.3292             B101        C  
742      2  PC 17608  262.3750  B57 B59 B63 B66        C

Категоризируйте возраст пассажиров "Титаника" на три группы:

1. **Категория 1**: Молодые пассажиры (до 30 лет)
   - Если возраст меньше 30, функция возвращает 1.

2. **Категория 2**: Пассажиры среднего возраста (от 30 до 54 лет)
   - Если возраст от 30 до 54 включительно, функция возвращает 2.

3. **Категория 3**: Пожилые пассажиры (55 лет и старше)
   - Для всех остальных случаев (то есть 55 лет и старше) функция возвращает 3.

In [8]:
def categorized_passengers_for_age(age):
  if pd.isna(age):  # Проверка на пропущенные значения
    return None
  if age < 30:
    return 1
  elif age >= 30 and age < 54:
    return 2
  elif age >= 55:
    return 3

df['age_category'] = df['Age'].apply(categorized_passengers_for_age)

# Вывод первых строк для проверки
print(df[['Age', 'age_category']].head())


    Age  age_category
0  22.0           1.0
1  38.0           2.0
2  26.0           1.0
3  35.0           2.0
4  35.0           2.0


**1. Сколько мужчин / женщин находилось на борту?**
- 412 мужчин и 479 женщин
- 314 мужчин и 577 женщин
- 479 мужчин и 412 женщин
- 577 мужчин и 314 женщин

In [9]:
gender_counts = df['Sex'].value_counts()

print(gender_counts)

Sex
male      577
female    314
Name: count, dtype: int64


**2. Выведите распределение переменной `Pclass` (социально-экономический статус) и это же распределение, только для мужчин / женщин по отдельности. Сколько было мужчин 2-го класса?**
- 104
- 108
- 112
- 125

In [10]:
# Распределение Pclass для всех пассажиров
pclass_distribution = df['Pclass'].value_counts()
print("Распределение Pclass для всех пассажиров:")
print(pclass_distribution)

# Распределение Pclass для мужчин
male_pclass_distribution = df[df['Sex'] == 'male']['Pclass'].value_counts()
print("\nРаспределение Pclass для мужчин:")
print(male_pclass_distribution)

# Распределение Pclass для женщин
female_pclass_distribution = df[df['Sex'] == 'female']['Pclass'].value_counts()
print("\nРаспределение Pclass для женщин:")
print(female_pclass_distribution)

# Количество мужчин 2-го класса
men_second_class = male_pclass_distribution.get(2, 0)
print("\nКоличество мужчин 2-го класса:", men_second_class)

Распределение Pclass для всех пассажиров:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Распределение Pclass для мужчин:
Pclass
3    347
1    122
2    108
Name: count, dtype: int64

Распределение Pclass для женщин:
Pclass
3    144
1     94
2     76
Name: count, dtype: int64

Количество мужчин 2-го класса: 108


**3. Каковы медиана и стандартное отклонение платежей (`Fare`)? Округлите до 2 десятичных знаков.**
- Медиана – 14.45, стандартное отклонение – 49.69
- Медиана – 15.1, стандартное отклонение – 12.15
- Медиана – 13.15, стандартное отклонение – 35.3
- Медиана – 17.43, стандартное отклонение – 39.1

**4. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?**
- 22.7% среди молодых и 40.6% среди старых
- 40.6% среди молодых и 22.7% среди старых
- 35.3% среди молодых и 27.4% среди старых
- 27.4% среди молодых и 35.3% среди старых

**5. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?**
- 30.2% среди мужчин и 46.2% среди женщин
- 35.7% среди мужчин и 74.2% среди женщин
- 21.1% среди мужчин и 46.2% среди женщин
- 18.9% среди мужчин и 74.2% среди женщин

**6. Найдите самое популярное имя среди пассажиров Титаника мужского пола?**
- Charles
- Thomas
- William
- John

**7. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:**
- В среднем мужчины 1-го класса старше 40 лет
- В среднем женщины 1-го класса старше 40 лет
- Мужчины всех классов в среднем старше женщин того же класса
- В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса